### PR TITLE
LPS-134172

### PR DESCRIPTION
**Pull request by javierdearcos**
Add the support of nestedFields on Object API for one to many relationships.

Add the logic to DTO converter as we decided.

We always add the property {relationshipName}Id to the objectEntry, and add another property {relationshipName} if it is specified as a nested field.

An example with a "manager" relationship:
Captura de pantalla 2021-11-16 a las 14 37 36

Currently, we support the one-to-many relationship from the "many" side to the "one", as we don't have the information present in the other ObjectEntry to establish the relation "one to many". We can add a new service method to get all the related objects of an objectEntryId with this goal.

The actions block is set to null in the embedded object entry, as we will need a way to build a new URIInfo for the new ObjectEntry (using the ObjectDefinition.getRESTContextPath and the objectEntryId) and I didn't find a way. I think we can create new methods in ActionUtil to provide a Supplier<URI> to build the action URI instead of passing the URIInfo.

I wasn't able to test the many-to-many relationships from the UI to use an example about how to get it from the APIs so it is not covered on this PR

### CODE REVIEW

**Comment by nhpatt** [comment]
You can not generate directly the resource files, as you can not inject the resource directly (you need to inject the factory component). I prepared a PR about it (https://issues.liferay.com/browse/LPS-134176) but did not sent it, let me see if I can restore it.

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/jaxrs/writer/interceptor/ObjectNestedFieldsWriterInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+		_setResourceContexts(
+			nestedFieldsContext.getMessage(), objectEntryResource);
+
+		Long objectEntryId = _getRelatedObjectEntryId(fieldName, objectEntry);
+
+		if (objectEntryId == null) {
+			return null;
+		}
+
+		return objectEntryResource.getObjectEntry(objectEntryId);
```
Comment (nhpatt) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/jaxrs/writer/interceptor/ObjectNestedFieldsWriterInterceptor.java:139]: Do we need the resource here? How do we get an ObjectDefinition?

Comment (javierdearcos) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/jaxrs/writer/interceptor/ObjectNestedFieldsWriterInterceptor.java:139]: > You can not generate directly the resource files, as you can not inject the resource directly (you need to inject the factory component). I prepared a PR about it (https://issues.liferay.com/browse/LPS-134176) but did not sent it, let me see if I can restore it.

Removing the Prototype scope made it work, but I guess I didn't consider some other cases... The good news is that this is something we have in our radar to achieve and even better if you had something prepared for this

**Comment by javierdearcos** [comment]
>You can not generate directly the resource files, as you can not inject the resource directly (you need to inject the factory component). I prepared a PR about it (https://issues.liferay.com/browse/LPS-134176) but did not sent it, let me see if I can restore it.

Removing the Prototype scope made it work, but I guess I didn't consider some other cases... The good news is that this is something we have in our radar to achieve and even better if you had something prepared for this

**Comment by javierdearcos** [comment]
ci:test:sf

**Comment by javierdearcos** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/jaxrs/writer/interceptor/ObjectNestedFieldsWriterInterceptor.java (hunk 1 above)
Comment (javierdearcos) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/jaxrs/writer/interceptor/ObjectNestedFieldsWriterInterceptor.java]: Does not appl with the ObjectEntryDTO approach

**Comment by nhpatt** [comment]
> We can add a new service method to get all the related objects of an objectEntryId with this goal.
I think we should, in a later PR

**Comment by nhpatt** [comment]
> The actions block is set to null in the embedded object entry, as we will need a way to build a new URIInfo for the new ObjectEntry (using the ObjectDefinition.getRESTContextPath and the objectEntryId) and I didn't find a way. I think we can create new methods in ActionUtil to provide a Supplier to build the action URI instead of passing the URIInfo.

I don't consider a nested actions as something important but I guess that you can always create a new `UriInfo` as we do in import/export or expose more methods in actionUtil as you mentioned.

**Review by nhpatt — APPROVED** [review]

**Comment by nhpatt** [comment]
ci:forward

**Comment by javierdearcos** [comment]
ci:test:relevant

**Pull request closed without merging**